File: python/src/ozbayMetricCalculator.py
import editDistance
import jaccardDistance
import longestCommonSubstringDistance

__author__ = "Hakan Ozbay"
__maintainer__ = "Hakan Ozbay"
__version__ = "1.0.0"

def calculate(base,term):
	base = base.lower()
	term = term.lower()
	edit_distance = editDistance.calculate(base,term)
	jaccard_distance = jaccardDistance.calculate(base,term)
	lcss_distance = longestCommonSubstringDistance.calculate(base,term)

	jaccard_metric = jaccard_distance
	lcss_metric = lcss_distance/(len(base) if len(base) > len(term) else len(term))
	edit_distance_metric = edit_distance/len(base)
	
	if jaccard_metric == 1:
		metric = edit_distance
	elif jaccard_metric == 0:
		metric = edit_distance_metric
	else:
		metric = jaccard_metric * edit_distance

	if lcss_metric > 0:
		metric /= lcss_metric
	else:
		metric *= (len(base) if len(base) < len(term) else len(term))

	return metric

File: java/src/main/java/com/hakanozbay/metric/OzbayMetricCalculator.java
package com.hakanozbay.metric;

public class OzbayMetricCalculator implements Metric {

	public double calculate(String base, String term) {
		
		String baseLowerCase = base.toLowerCase();
		String termLowerCase = term.toLowerCase();
		
		double jaccardDistance = calculateJaccardDistannce(baseLowerCase, termLowerCase);
		double editDistance = calculateEditDistance(baseLowerCase, termLowerCase);
		double longestCommonSubstringDistance = calculateLongestCommonSubstringDistance(baseLowerCase, termLowerCase);
	
		double jaccardMetric = jaccardDistance;
		double lcssMetric = longestCommonSubstringDistance/(base.length() > term.length()? base.length() : term.length());
		double editDistanceMetric = editDistance/base.length();
		
		double metric;
		
		if (jaccardMetric == 1)
			metric = editDistance;
		else if (jaccardMetric == 0)
			metric = editDistanceMetric;
		else
			metric = jaccardMetric * editDistance;

		if (lcssMetric > 0)
			metric /= lcssMetric;
		else
			metric *= (base.length() < term.length()? base.length(): term.length());

		return metric;
	}


	protected double calculateJaccardDistannce(String base, String term)
	{
		return new JaccardDistance().calculate(base, term);
	}
	
	protected double calculateEditDistance(String base, String term) 
	{
		return new EditDistance().calculate(base, term);
	}

	protected double calculateLongestCommonSubstringDistance(String baseLowerCase, String termLowerCase) 
	{
		return new LongestCommonSubstringDistance().calculate(baseLowerCase, termLowerCase);
	}
	
}
